lib/spandex_ecto/ecto_logger.ex

```diff
@@ -18,75 +24,107 @@ defmodule SpandexEcto.EctoLogger do
+      completion_time: now,
+      service: service,
+      resource: query,
+      type: :db,
+      sql_query: [
+        query: query,
+        rows: inspect(num_rows),
+        db: database
+      ],
+      tags: tags(log_entry)
+    )
+
+    report_error(tracer, log_entry)
+
+    if queue_time != 0 do
+      tracer.start_span("queue")
+      tracer.update_span(service: service, start: start, completion_time: start + queue_time)
+      tracer.finish_span()
+    end
 
-      start = now - (queue_time + query_time + decoding_time)
+    if query_time != 0 do
+      tracer.start_span("run_query")
 
-      tracer.start_span(
-        "query",
-        start: start,
-        completion_time: now,
+      tracer.update_span(
         service: service,
-        resource: query,
-        type: :db,
-        sql_query: [
-          query: query,
-          rows: inspect(num_rows),
-          db: database
-        ],
-        tags: tags(log_entry)
+        start: start + queue_time,
+        completion_time: start + queue_time + query_time
       )
 
-      Logger.metadata(trace_id: tracer.current_trace_id(), span_id: tracer.current_span_id())
+      tracer.finish_span()
+    end
 
-      report_error(tracer, log_entry)
+    if decoding_time != 0 do
+      tracer.start_span("decode")
 
-      if queue_time != 0 do
-        tracer.start_span("queue")
-        tracer.update_span(service: service, start: start, completion_time: start + queue_time)
-        tracer.finish_span()
-      end
+      tracer.update_span(
+        service: service,
+        start: start + queue_time + query_time,
+        completion_time: now
+      )
+
+      tracer.finish_span()
+    end
 
-      if query_time != 0 do
-        tracer.start_span("run_query")
+    finish_ecto_trace(caller_pid, tracer)
 
-        tracer.update_span(
-          service: service,
-          start: start + queue_time,
-          completion_time: start + queue_time + query_time
-        )
+    log_entry
+  end
 
-        tracer.finish_span()
-      end
+  defp finish_ecto_trace(caller_pid, tracer) do
+    if caller_pid != self() do
+      tracer.finish_trace()
+    else
+      tracer.finish_span()
+    end
+  end
 
-      if decoding_time != 0 do
-        tracer.start_span("decode")
+  defp setup(caller_pid, tracer) when is_pid(caller_pid) do
+    if caller_pid == self() do
+      tracer.start_span("query")
+    else
+      case Process.info(caller_pid)[:dictionary][:spandex_trace] do
+        nil ->
+          tracer.start_trace("query")
 
-        tracer.update_span(
-          service: service,
-          start: start + queue_time + query_time,
-          completion_time: now
-        )
+        %Trace{id: trace_id, stack: [%Span{id: span_id} | _]} ->
+          tracer.continue_trace("query", %SpanContext{trace_id: trace_id, parent_id: span_id})
 
-        tracer.finish_span()
+        %Trace{id: trace_id, stack: []} ->
+          tracer.continue_trace("query", %SpanContext{trace_id: trace_id})
```

Comment:
I am curious how would flame graph look in this case (i.e. is this new span going to be attached to the top level trace? if so, it is incorrect)

Reply:
Yes, that's correct. But they had here, before they removed whole preload support https://github.com/spandex-project/spandex_ecto/blob/18d5f34158b155d10b18759082b793edb9e06ae0/lib/spandex_ecto/ecto_logger.ex#L98